File: go/server/network_check.go
package server

import (
	"fmt"
	"strings"

	"github.com/solana-foundation/mpp-sdk/go"
)

// SurfpoolBlockhashPrefix is the base58 prefix embedded in every blockhash
// returned by the Surfpool localnet implementation. Servers configured for
// any network OTHER than localnet use this prefix to detect wrong-RPC
// client mistakes.
const SurfpoolBlockhashPrefix = "SURFNETxSAFEHASH"

// LocalnetNetwork is the network slug for Solana's local validator. The
// only network for which a Surfpool-prefixed blockhash is valid.
const LocalnetNetwork = "localnet"

// CheckNetworkBlockhash is an asymmetric check: it rejects credentials
// whose signed blockhash carries the Surfpool prefix when the server is
// configured for any network other than `localnet`.
//
// Returns nil in every other case — a non-Surfpool blockhash is
// undetectable as wrong-cluster from the slug alone, so we let the
// downstream broadcast handle it.
func CheckNetworkBlockhash(network, blockhashB58 string) error {
	if !strings.HasPrefix(blockhashB58, SurfpoolBlockhashPrefix) {
		return nil
	}
	if network == LocalnetNetwork {
		return nil
	}
	_ = blockhashB58 // intentionally unused: blockhash detail is debug-grade,
	// not actionable for end users — keep the message terse.
	return mpp.NewError(mpp.ErrCodeWrongNetwork, fmt.Sprintf(
		"Signed against localnet but the server expects %s. "+
			"Switch your client RPC to %s and re-sign.",
		network, network,
	))
}

File: lua/mpp/server/network_check.lua
--[[
Network / blockhash sanity check.

The Surfpool localnet implementation embeds a recognizable prefix into
every blockhash it returns. We use this to catch the common footgun where
a client signs a transaction against a Surfpool RPC and submits it to a
server configured for a real cluster (mainnet/devnet).

The check is asymmetric:

  - If the blockhash starts with the Surfpool prefix, the transaction
    was DEFINITELY signed against a Surfpool localnet. The only network
    slug for which that's valid is `localnet` — any other slug must
    reject the credential up-front.

  - If the blockhash does NOT start with the Surfpool prefix, we can't
    tell what cluster it came from (real localnet doesn't add a prefix
    either), so we accept it.

The Lua server delegates transaction decoding to user-provided
`verify_payment` callbacks (since Lua has no built-in Solana SDK), so this
module exposes a pure check that callback authors must invoke once they
have the blockhash in hand.
]]

local M = {}

--- Base58 prefix embedded in every blockhash returned by the Surfpool
--- localnet implementation.
M.SURFPOOL_BLOCKHASH_PREFIX = 'SURFNETxSAFEHASH'

--- Network slug for Solana's local validator. The only network for which
--- a Surfpool-prefixed blockhash is valid.
M.LOCALNET_NETWORK = 'localnet'

--- Pure check: returns nil if the blockhash is acceptable for the given
--- network, or an error table `{ code = 'wrong-network', message = '...' }`
--- if a Surfpool-prefixed blockhash is being submitted to a server
--- configured for any network other than `localnet`.
---
--- A non-Surfpool blockhash is accepted on any network because we can't
--- tell which real cluster it came from.
---
--- @param network string Server-configured network slug.
--- @param blockhash_b58 string Base58-encoded recent blockhash from the signed tx.
--- @return table|nil err Error table on mismatch, nil on success.
function M.check_network_blockhash(network, blockhash_b58)
  local hash = blockhash_b58 or ''
  local has_prefix = string.sub(hash, 1, #M.SURFPOOL_BLOCKHASH_PREFIX) == M.SURFPOOL_BLOCKHASH_PREFIX
  if not has_prefix then
    return nil
  end
  if network == M.LOCALNET_NETWORK then
    return nil
  end
  -- Blockhash detail is debug-grade and not actionable for end users.
  return {
    code = 'wrong-network',
    message = 'Signed against localnet but the server expects ' .. tostring(network) .. '. '
      .. 'Switch your client RPC to ' .. tostring(network) .. ' and re-sign.',
  }
end

--- Convenience: same as `check_network_blockhash` but raises via `error()`
--- instead of returning the error table. Designed for use inside
--- `verify_payment` callbacks that follow the rest-of-server convention
--- of raising on validation failure.
function M.assert_network_blockhash(network, blockhash_b58)
  local err = M.check_network_blockhash(network, blockhash_b58)
  if err then
    error(err.message)
  end
end

return M
